File: LeetCode/01578. Minimum Time to Make Rope Colorful/01578.go
// https://leetcode.com/problems/minimum-time-to-make-rope-colorful

func minCost(colors string, neededTime []int) (ans int) {
    var (
        sumTime, maxTime int
        prevColor byte
    )

    for i, size := 0, len(colors); i < size; i++ {
        if prevColor == colors[i] {
            sumTime += neededTime[i]
            maxTime = max(maxTime, neededTime[i])
        } else {
            ans += sumTime - maxTime

            prevColor = colors[i]
            sumTime = neededTime[i]
            maxTime = neededTime[i]
        }
    }

    return ans + sumTime - maxTime
}

File: LeetCode/01578. Minimum Time to Make Rope Colorful/01578.c
int minCost(char * colors, int *neededTime, int neededTimeSize) {
    int count = 0;
    int i, head, maxOfSame, sumOfSame;
    
    for (i = 1, head = 0, maxOfSame = sumOfSame = neededTime[0]; i < neededTimeSize; ++i) {
        if (colors[i-1] == colors[i]) {
            maxOfSame = (maxOfSame > neededTime[i]) ? maxOfSame : neededTime[i];
            sumOfSame += neededTime[i];
        } else {
            if (i-1 > head) {
                count += sumOfSame - maxOfSame;
            }
            
            head = i;
            maxOfSame = sumOfSame = neededTime[i];
        }
    }
    
    if (i-1 > head) {
        count += sumOfSame - maxOfSame;
    }
    
    return count;
}
